lib/Sema/CSSimplify.cpp

```diff
@@ -425,6 +445,14 @@ matchCallArguments(ArrayRef<AnyFunctionType::Param> args,
 
   // If we have a trailing closure, it maps to the last parameter.
   if (hasTrailingClosure && numParams > 0) {
+    // If there is no suitable last parameter to accept the trailing closure,
+    // notify the listener and bail if we need to.
+    if (!acceptsTrailingClosure(params[numParams - 1])) {
+      if (listener.trailingClosureMismatch(numParams - 1, numArgs - 1))
```

Comment:
In the spirit of `matchCallArguments` used only to match parameters to arguments based on labels and positions, it seems that better place for this logic could be in other overload of `matchCallArguments` where types are matched, the problem with that though is that we currently don't diagnose missing parameters via new diagnostic framework, that would have made diagnosing trailing closures much easier...